### PR TITLE
Python Cloud Object Storage Support

### PR DESCRIPTION
**Pull request by bassel-zeidan**
**Background**

This PR adds a class which gives the ability to setup the hadoop configs to connect to the cloud object storage (COS).

### CODE REVIEW

**Comment by gadamc** [comment]
Remove the version bump from this commit. I bump the version when I make a release (https://github.com/gadamc/release-python)

**Comment by gadamc** [comment]
I think it will be good practice going forward that you post a link here directly to the tests that demonstrate it's working. That would be the notebook link you sent to me on Slack. Remember to use the #@hidden_cell magic to hide your credentials thought

**Comment by gadamc** [comment]
You need to add instructions and examples to the README. Also, you need to describe the difference between 'CloudObjectStorage' and the bluemix/softlayer object storage

**Comment by gadamc** [comment]
I'm not a fan of the CamelCase class 'CloudObjectStorage'. At the same time, I don't like 'cloudobjectstorage' or 'cloud_object_storage'.

**Comment by gilv** [comment]
@gadamc @bassel-zeidan I will review it. Some things not clear here, will comment later.

File: python/ibmos2spark/osconfig.py [hunk 1]
```diff
@@ -170,3 +170,48 @@ def __init__(self, sparkcontext, credentials, name=None, public=False, swift2d_d
...
+
+        for i in range(len(credential_key_list)):
+            key = credential_key_list[i]
+            if (not key in credentials):
+                raise ValueError("Invalid input: credentials.{} is required!".format(key))
+
+        # setup config
+        prefix = "fs.s3d.service"
+        hconf = sparkcontext._jsc.hadoopConfiguration()
+        hconf.set(prefix + ".impl", driver)
```
Comment (gilv) [inline python/ibmos2spark/osconfig.py]: @bassel-zeidan This key suppose to be already setup in core-sites.xml. There is no need to setup it again.

**Comment by gilv** [comment]
@bassel-zeidan why we need this class? why not a simple function?

**Comment by gadamc** [comment]
@gilv @bassel-zeidan -- the reason Bassel has added this as a class is because he's following the precedent I set when I first created this project. 

I have separate classes, mostly for historical reasons, in order to support previous 'swift' and 'swift2d' protocols. But also, having these as classes that a user instantiates facilitates, slightly, using multiple different object storages simultaneously because the class will build the swift URL based on the protocol, and configuration name. So, it's meant to facilitate users experience. Perhaps there could have been a better design, but when I first built this I didn't actually expect anybody buy myself to do use it. :)

**Comment by bassel-zeidan** [comment]
@gadamc  thanks for taking a look.

1) version upgrade removed. 
2) regarding testing:
```
https://apsportal.ibm.com/analytics/notebooks/628a1960-6f48-4dc3-b57b-5666363831d9/view?access_token=4ada3a91af5a4e11c0ee47b7c0b3f18a1623ba70075d0a0ab26443ad713f33cf
```

3) I updated the Readme file and add an example on how to use COS.

**Comment by bassel-zeidan** [comment]
@gilv thanks for the feedback.

regarding having it as a class, as Adam mentioned this is convention that has been followed before and i think it makes sense because normally you define your configurations once (make an instance of the class) and try to access files multiple times (using url function) and such a structure helps with that.

**Comment by bassel-zeidan** [comment]
@gilv thanks for the feedback.

regarding having it as a class, as Adam mentioned this is convention that has been followed before and i think it makes sense because normally you define your configurations once (make an instance of the class) and try to access files multiple times (using url function) and such a structure helps with that.

**Comment by bassel-zeidan** [comment]
In case you have no more concerns, let's get it merged.

**Comment by bassel-zeidan** [comment]
@gilv @gadamc I was just wondering, should we change the pattern for cos class and write it as a function instead. The thing is, with the current implementation if you instantiate two instances from this class, you will overwrite the hadoop configs with the second call (which can be a bit confusing if the user is not familiar with the internal implementation). In Bluemix and Softlayer classes it was possible to define multiple configs by defining an extra name but not with COS.

Would it make sense to define a function instead of a class that return the url function which can be used later to access files. i.e.
```
def cos_connection(credentials, .....):
            #set hadoop config
            hconf = ...
            ....
            
            # define url function that will be used multiple time
            def url(bucket_name, file):
                  return .....

            return url
```

it can be then used as (common case: cos_connection is called once and url function is called multiple times):
```
cos_url_func = cos_connection(credentials, ..)
cos_url_func(bucket_name, file1_name)
...
...
cos_url_func(bucket_name, file2_name)

```

What are you thoughts on that?

File: python/ibmos2spark/osconfig.py [hunk 2]
```diff
@@ -170,3 +170,68 @@ def __init__(self, sparkcontext, credentials, name=None, public=False, swift2d_d
...
+        # check if all required values are availble
+        credential_key_list = ["endpoint", "access_key", "secret_key"]
+
+        for i in range(len(credential_key_list)):
+            key = credential_key_list[i]
+            if (not key in credentials):
+                raise ValueError("Invalid input: credentials.{} is required!".format(key))
+
+        # setup config
+        prefix = "fs.s3d.service"
```
Comment (gadamc) [inline python/ibmos2spark/osconfig.py]: Can we make this 'fs.s3d.service.' + cos_id?

**Comment by gadamc** [comment]
@bassel-zeidan @gilv 

In the 'softlayer' and 'bluemix' classes, we append a configuration name to the "prefix". 
https://github.com/ibm-watson-data-lab/ibmos2spark/blob/PythonCOSSupport/python/ibmos2spark/osconfig.py#L158

This allows a user to create and use multiple softlayer / bluemix object storage accounts. For example, if I need to read data from one account and then write to another. I think this would be especially important to be able to define multiple configurations in a parallelized process where each worker node is interacting with the COS. Right? 

Can we do this for COS? That is, can we change line 219?   That would allow a person with multiple accounts configure access to them independently with the same code.

**Comment by bassel-zeidan** [comment]
@gadamc unfortunately that is not possible. I tried to do that in different ways but it seems that it is not supported.

File: python/ibmos2spark/osconfig.py (hunk 2 above)
Comment (bassel-zeidan) [inline python/ibmos2spark/osconfig.py]: Tried it. Didn't work out unfortunately

**Comment by bassel-zeidan** [comment]
@gilv do you know of a way to support connecting to multiple COSs? I tried the same approach that has been followed in bluemix and softlayer classes (appending unique OS id to the prefix of the settings e.g. `"fs.swift2d.service." + name`). This doesn't seem to work with COS. Any ideas?

**Comment by gilv** [comment]
@bassel-zeidan It was resolved actually. You can setup different service names for COS, similar to Swift. But it's not yet deployed to Spark as a Service. This commit https://github.ibm.com/cloud-platforms/stocator-s3/commit/519bf601e6d934f7c69de9739e4ea6afe6dc9c68 will part of 1.4 version that we will release in a week or so.

**Comment by bassel-zeidan** [comment]
Great! I think we should utilise this update in ibmos2spark implementation and support defining multiple configs.  However, It might take some time until the new version of Stocator is released and until it gets deployed on Spark service. I would suggest that we introduce COS update to ibmos2spark (Python, Scala and R) and make a new release then we can support the multiple configs support in the next version. I will open a new item for that and assign it to myself.

@gadamc @gilv Do you agree to that?

**Comment by bassel-zeidan** [comment]
tests on the latest version of this PR:

https://apsportal.ibm.com/analytics/notebooks/b7a32505-8df5-49c4-a230-1dee851f0e03/view?access_token=8941ac5d33113abf988298a5c86c56c4d260510faa39f097150142a7098e6eb5

**Comment by gadamc** [comment]
@bassel-zeidan I'd rather wait. If @gilv thinks its only a week or two away, I don't see the rush. 
I'd rather have the COS usage consistent with the other classes -- which includes supplying a 'configuration_name'.  And I'd also not want to release twice. @gilv -- how confident are you in the timeline for the update to support a multiple configurations?

**Comment by bassel-zeidan** [comment]
As discussed, i am merging this PR.

**Pull request merged**